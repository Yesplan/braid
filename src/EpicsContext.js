import { createContext } from 'react';

const EpicsContext = createContext({});

export { EpicsContext };
