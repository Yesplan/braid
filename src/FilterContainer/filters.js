export const hasUnresolvedBlockers = blockers =>
  blockers.some(blocker => !blocker.resolved);

const isBlocked = story => hasUnresolvedBlockers(story.blockers);

export const filterByOwner = (story, ownerIds) => {
  if (ownerIds.length === 0) {
    return true;
  }

  return story.ownerIds.some(id => ownerIds.includes(id));
};

export const filterByEpic = (story, epicIds) => {
  if (epicIds.length === 0) {
    return true;
  }

  return story.labels.some(label => epicIds.includes(label.id));
};

export const filterByType = (story, typeNames) => {
  if (typeNames.length === 0) {
    return true;
  }

  return (
    typeNames.includes(story.storyType) ||
    (typeNames.includes('blocked') && isBlocked(story))
  );
};

export const filterByStoryStates = (story, storyStates) =>
  storyStates.includes(story.currentState);
