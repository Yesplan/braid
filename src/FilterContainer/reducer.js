import { arrayToggle, arrayRotateForward, arrayRotateBackward } from '../utils';

export const toggleOwner = 'TOGGLE_OWNER';
export const clearOwners = 'CLEAR_OWNERS';
export const selectNextOwner = 'SELECT_NEXT_OWNER';
export const selectPrevOwner = 'SELECT_PREV_OWNER';
export const toggleType = 'TOGGLE_TYPE';

export const reducer = (state, action, uniqueOwnerIds) => {
  const { type, payload } = action;

  switch (type) {
    case toggleOwner: {
      return {
        ...state,
        selectedOwners: arrayToggle(state.selectedOwners, payload)
      };
    }

    case clearOwners: {
      return {
        ...state,
        selectedOwners: []
      };
    }

    case selectNextOwner: {
      return {
        ...state,
        selectedOwners: [
          arrayRotateForward(uniqueOwnerIds, state.selectedOwners[0])
        ]
      };
    }

    case selectPrevOwner: {
      return {
        ...state,
        selectedOwners: [
          arrayRotateBackward(uniqueOwnerIds, state.selectedOwners[0])
        ]
      };
    }

    case toggleType: {
      return {
        ...state,
        selectedTypes: arrayToggle(state.selectedTypes, payload)
      };
    }

    default: {
      return { ...state };
    }
  }
};
