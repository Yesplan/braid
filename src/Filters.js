import React, { useContext, Fragment } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faBackward,
  faForward,
  faTimes
} from '@fortawesome/free-solid-svg-icons';
import { Initials } from './Initials';
import {
  FilterContext,
  toggleOwner,
  clearOwners,
  selectNextOwner,
  selectPrevOwner,
  toggleEpic,
  clearEpics,
  selectNextEpic,
  selectPrevEpic,
  toggleType
} from './FilterContainer';
import { storyTypes } from './storyTypes';
import { EpicName } from './EpicName';

const Filters = () => {
  const {
    selectedTypes,
    selectedOwners,
    selectedEpics,
    uniqueOwnerIds,
    uniqueEpicIds,
    dispatch
  } = useContext(FilterContext);

  const handleOwnerClick = event => {
    event.target.focus();
    dispatch({
      type: toggleOwner,
      payload: Number(event.currentTarget.value)
    });
  };

  const handleEpicClick = event => {
    event.target.focus();
    dispatch({
      type: toggleEpic,
      payload: Number(event.currentTarget.value)
    });
  };

  const handleTypeClick = event => {
    event.target.focus();
    dispatch({ type: toggleType, payload: event.currentTarget.value });
  };

  const handleNextOwnerClick = event => {
    event.target.focus();
    dispatch({ type: selectNextOwner });
  };

  const handlePrevOwnerClick = event => {
    event.target.focus();
    dispatch({ type: selectPrevOwner });
  };

  const handleClearOwners = event => {
    event.target.focus();
    dispatch({ type: clearOwners });
  };

  const handleNextEpicClick = event => {
    event.target.focus();
    dispatch({ type: selectNextEpic });
  };

  const handlePrevEpicClick = event => {
    event.target.focus();
    dispatch({ type: selectPrevEpic });
  };

  const handleClearEpics = event => {
    event.target.focus();
    dispatch({ type: clearEpics });
  };

  return (
    <Fragment>
      <div className="media">
        <div className="media-left">
          <h5 className="subtitle is-5" style={{ lineHeight: '1.8' }}>
            By owner
          </h5>
        </div>
        <div className="media-content">
          <div className="field is-grouped is-grouped-multiline">
            {uniqueOwnerIds.map(id => (
              <p className="control" key={id}>
                <button
                  key={id}
                  value={id}
                  className={`button is-rounded ${
                    selectedOwners.indexOf(id) !== -1 ? 'is-primary' : ''
                  }`}
                  onClick={handleOwnerClick}
                >
                  <Initials id={id} />
                </button>
              </p>
            ))}

            <div
              className="field has-addons"
              style={{ marginBottom: '.75rem' }}
            >
              <p className="control">
                <button
                  className="button"
                  data-testid="prevOwner"
                  onClick={handlePrevOwnerClick}
                >
                  <span className="icon">
                    <FontAwesomeIcon icon={faBackward} />
                  </span>
                  <span>
                    <span className="is-underline">P</span>rev
                  </span>
                </button>
              </p>
              <p className="control">
                <button
                  className="button"
                  data-testid="nextOwner"
                  onClick={handleNextOwnerClick}
                >
                  <span className="icon">
                    <FontAwesomeIcon icon={faForward} />
                  </span>
                  <span>
                    <span className="is-underline">N</span>ext
                  </span>
                </button>
              </p>

              <p className="control">
                <button
                  className="button"
                  data-testid="clearOwners"
                  onClick={handleClearOwners}
                  disabled={selectedOwners.length === 0}
                >
                  <span className="icon">
                    <FontAwesomeIcon icon={faTimes} />
                  </span>
                  <span>
                    <span className="is-underline">C</span>lear
                  </span>
                </button>
              </p>
            </div>
          </div>
        </div>
      </div>
      <div className="media is-marginless" style={{ border: '0' }}>
        <div className="media-left">
          <h5 className="subtitle is-5" style={{ lineHeight: '1.8' }}>
            By type
          </h5>
        </div>
        <div className="media-content">
          <div className="buttons">
            {storyTypes.map(({ key, activeClass }) => (
              <button
                value={key}
                key={key}
                className={`button is-rounded is-uppercase ${
                  selectedTypes.includes(key) ? activeClass : ''
                }`}
                onClick={handleTypeClick}
              >
                {key}
              </button>
            ))}
          </div>
        </div>
      </div>
      <div className="media is-marginless" style={{ border: '0' }}>
        <div className="media-left">
          <h5 className="subtitle is-5" style={{ lineHeight: '1.8' }}>
            By epic
          </h5>
        </div>
        <div className="media-content">
          <div className="field is-grouped is-grouped-multiline">
            {uniqueEpicIds.map(id => (
              <p className="control" key={id}>
                <button
                  key={id}
                  value={id}
                  className={`button is-rounded ${
                    selectedEpics.indexOf(id) !== -1 ? 'is-primary' : ''
                  }`}
                  onClick={handleEpicClick}
                >
                  <EpicName id={id} />
                </button>
              </p>
            ))}

            <div
              className="field has-addons"
              style={{ marginBottom: '.75rem' }}
            >
              <p className="control">
                <button
                  className="button"
                  data-testid="prevEpic"
                  onClick={handlePrevEpicClick}
                >
                  <span className="icon">
                    <FontAwesomeIcon icon={faBackward} />
                  </span>
                  <span>
                    <span className="is-underline">P</span>rev
                  </span>
                </button>
              </p>
              <p className="control">
                <button
                  className="button"
                  data-testid="nextEpic"
                  onClick={handleNextEpicClick}
                >
                  <span className="icon">
                    <FontAwesomeIcon icon={faForward} />
                  </span>
                  <span>
                    <span className="is-underline">N</span>ext
                  </span>
                </button>
              </p>

              <p className="control">
                <button
                  className="button"
                  data-testid="clearEpics"
                  onClick={handleClearEpics}
                  disabled={selectedEpics.length === 0}
                >
                  <span className="icon">
                    <FontAwesomeIcon icon={faTimes} />
                  </span>
                  <span>
                    <span className="is-underline">C</span>lear
                  </span>
                </button>
              </p>
            </div>
          </div>
        </div>
      </div>
    </Fragment>
  );
};

export { Filters };
