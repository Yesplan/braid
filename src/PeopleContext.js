import { createContext } from 'react';

const { Consumer, Provider } = createContext({});

export { Consumer as PeopleConsumer, Provider as PeopleProvider };
