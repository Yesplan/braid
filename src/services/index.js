export { default as window } from './window';
export * from './localStorage';
