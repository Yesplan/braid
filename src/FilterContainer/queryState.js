import { window } from '../services';
import { normalizeOwners, normalizeTypes } from './normalize';

const queryString = require('query-string');

const queryStringOptions = {
  arrayFormat: 'comma'
};

function getQueryState(uniqueOwnerIds) {
  const queryState = queryString.parse(
    window.location.search,
    queryStringOptions
  );

  return {
    selectedOwners: normalizeOwners(queryState.selectedOwners, uniqueOwnerIds),
    selectedTypes: normalizeTypes(queryState.selectedTypes)
  };
}

function setQueryState(state) {
  window.history.pushState(
    {},
    document.title,
    `${window.location.pathname}?${queryString.stringify(
      state,
      queryStringOptions
    )}`
  );
}

export { getQueryState, setQueryState };
