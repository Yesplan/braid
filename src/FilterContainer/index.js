import React, { useState, createContext, useEffect, useMemo } from 'react';
import { arrayOf, number, node } from 'prop-types';
import { window } from '../services';
import { filterByOwner, filterByType, filterByStoryStates } from './filters';
import {
  reducer,
  selectNextOwner,
  selectPrevOwner,
  clearOwners
} from './reducer';
import { getQueryState, setQueryState } from './queryState';
import { useKeyup } from '../useKeyup';

const FilterContext = createContext();

const FilterContainer = ({ uniqueOwnerIds, children }) => {
  const [state, setState] = useState(getQueryState(uniqueOwnerIds));

  const dispatch = useMemo(
    () => action => {
      const nextState = reducer(state, action, uniqueOwnerIds);

      setState(nextState);
      setQueryState(nextState);
    },
    [state, uniqueOwnerIds]
  );

  useEffect(() => {
    const syncStates = () => {
      setState(getQueryState(uniqueOwnerIds));
    };

    window.addEventListener('popstate', syncStates);

    return () => {
      window.removeEventListener('popstate', syncStates);
    };
  }, [uniqueOwnerIds]);

  const keyMap = useMemo(
    () => ({
      n: () => dispatch({ type: selectNextOwner }),
      p: () => dispatch({ type: selectPrevOwner }),
      c: () => dispatch({ type: clearOwners })
    }),
    [dispatch]
  );

  useKeyup(keyMap);

  const filter = useMemo(
    () => (storyIds, stories, storyStates) =>
      storyIds
        .map(id => stories[id])
        .filter(
          story =>
            filterByOwner(story, state.selectedOwners) &&
            filterByType(story, state.selectedTypes) &&
            filterByStoryStates(story, storyStates)
        ),
    [state.selectedOwners, state.selectedTypes]
  );

  const value = useMemo(
    () => ({
      ...state,
      uniqueOwnerIds,
      dispatch,
      filter
    }),
    [state, uniqueOwnerIds, dispatch, filter]
  );

  return (
    <FilterContext.Provider value={value}>{children}</FilterContext.Provider>
  );
};

FilterContainer.propTypes = {
  uniqueOwnerIds: arrayOf(number),
  children: node
};

export { FilterContainer, FilterContext };
export { hasUnresolvedBlockers } from './filters';
export {
  toggleOwner,
  clearOwners,
  selectNextOwner,
  selectPrevOwner,
  toggleType
} from './reducer';
