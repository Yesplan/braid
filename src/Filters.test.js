import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Filters } from './Filters';
import { PeopleContext } from './PeopleContext';
import { EpicsContext } from './EpicsContext';
import { FilterContainer } from './FilterContainer';
import { window } from './services';

jest.mock('./services/window');

describe('Filters', () => {
  afterEach(() => {
    window.cleanup();
  });

  const people = {
    1: { initials: 'AA' },
    2: { initials: 'BB' },
    3: { initials: 'CC' }
  };

  const epics = {
    4: { name: 'E:AA' },
    5: { name: 'E:BB' },
    6: { name: 'E:CC' }
  };

  const uniqueOwnerIds = [1, 2, 3];
  const uniqueEpicIds = [4, 5, 6];

  const renderSubject = () =>
    render(
      <PeopleContext.Provider value={people}>
        <EpicsContext.Provider value={epics}>
          <FilterContainer
            uniqueOwnerIds={uniqueOwnerIds}
            uniqueEpicIds={uniqueEpicIds}
          >
            <Filters />
          </FilterContainer>
        </EpicsContext.Provider>
      </PeopleContext.Provider>
    );

  it('renders as per snapthot', () => {
    const { container } = renderSubject();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('reads state from window.location.search', () => {
    window.location.search =
      'selectedOwners=2,3&selectedEpics=5,6&selectedTypes=bug,feature';

    const { getByText } = renderSubject();

    expect(getByText('AA').parentElement).not.toHaveClass('is-primary');
    expect(getByText('BB').parentElement).toHaveClass('is-primary');
    expect(getByText('CC').parentElement).toHaveClass('is-primary');

    expect(getByText('E:AA').parentElement).not.toHaveClass('is-primary');
    expect(getByText('E:BB').parentElement).toHaveClass('is-primary');
    expect(getByText('E:CC').parentElement).toHaveClass('is-primary');

    expect(getByText('chore')).not.toHaveClass('is-info');
    expect(getByText('blocked')).not.toHaveClass('is-warning');
    expect(getByText('bug')).toHaveClass('is-danger');
    expect(getByText('feature')).toHaveClass('is-primary');
  });

  describe('owners', () => {
    it('toggles owner on click', () => {
      const { getByText } = renderSubject();

      expect(getByText('AA').parentElement).not.toHaveClass('is-primary');

      fireEvent.click(getByText('AA'));

      expect(getByText('AA').parentElement).toHaveClass('is-primary');
      expect(window.location.search).toEqual('selectedOwners=1');
    });
  });

  describe('next owner', () => {
    it('selects next owner on click', () => {
      window.location.search = 'selectedOwners=2';
      const { getByText, getByTestId } = renderSubject();

      expect(getByText('BB').parentElement).toHaveClass('is-primary');

      fireEvent.click(getByTestId('nextOwner'));

      expect(getByText('BB').parentElement).not.toHaveClass('is-primary');
      expect(getByText('CC').parentElement).toHaveClass('is-primary');
      expect(window.location.search).toEqual('selectedOwners=3');
    });
  });

  describe('previous owner', () => {
    it('triggers selectPrevOwner on click', () => {
      window.location.search = 'selectedOwners=2';

      const { getByText, getByTestId } = renderSubject();

      expect(getByText('BB').parentElement).toHaveClass('is-primary');

      fireEvent.click(getByTestId('prevOwner'));

      expect(getByText('BB').parentElement).not.toHaveClass('is-primary');
      expect(getByText('AA').parentElement).toHaveClass('is-primary');
      expect(window.location.search).toEqual('selectedOwners=1');
    });
  });

  describe('clear owners', () => {
    it('is disabled when no owners is selected', () => {
      const { getByTestId } = renderSubject();

      expect(getByTestId('clearOwners')).toHaveAttribute('disabled');
    });

    it('clears owners on click', () => {
      const { getByText, getByTestId } = renderSubject({
        selectedOwners: [1, 2]
      });

      fireEvent.click(getByTestId('clearOwners'));

      expect(getByText('AA').parentElement).not.toHaveClass('is-primary');
      expect(getByText('BB').parentElement).not.toHaveClass('is-primary');
      expect(window.location.search).toEqual('');
    });
  });

  describe('epics', () => {
    it('toggles epic on click', () => {
      const { getByText } = renderSubject();

      expect(getByText('E:AA').parentElement).not.toHaveClass('is-primary');

      fireEvent.click(getByText('E:AA'));

      expect(getByText('E:AA').parentElement).toHaveClass('is-primary');
      expect(window.location.search).toEqual('selectedEpics=4');
    });
  });

  describe('next epic', () => {
    it('selects next epic on click', () => {
      window.location.search = 'selectedEpics=5';
      const { getByText, getByTestId } = renderSubject();

      expect(getByText('E:BB').parentElement).toHaveClass('is-primary');

      fireEvent.click(getByTestId('nextEpic'));

      expect(getByText('E:BB').parentElement).not.toHaveClass('is-primary');
      expect(getByText('E:CC').parentElement).toHaveClass('is-primary');
      expect(window.location.search).toEqual('selectedEpics=6');
    });
  });

  describe('previous epic', () => {
    it('triggers selectPrevEpic on click', () => {
      window.location.search = 'selectedEpics=5';

      const { getByText, getByTestId } = renderSubject();

      expect(getByText('E:BB').parentElement).toHaveClass('is-primary');

      fireEvent.click(getByTestId('prevEpic'));

      expect(getByText('E:BB').parentElement).not.toHaveClass('is-primary');
      expect(getByText('E:AA').parentElement).toHaveClass('is-primary');
      expect(window.location.search).toEqual('selectedEpics=4');
    });
  });

  describe('clear epics', () => {
    it('is disabled when no epic is selected', () => {
      const { getByTestId } = renderSubject();

      expect(getByTestId('clearEpics')).toHaveAttribute('disabled');
    });

    it('clears epics on click', () => {
      const { getByText, getByTestId } = renderSubject({
        selectedEpics: [4, 5]
      });

      fireEvent.click(getByTestId('clearEpics'));

      expect(getByText('E:AA').parentElement).not.toHaveClass('is-primary');
      expect(getByText('E:BB').parentElement).not.toHaveClass('is-primary');
      expect(window.location.search).toEqual('');
    });
  });

  describe('types', () => {
    it('toggles bug on click', () => {
      const { getByText } = renderSubject();

      fireEvent.click(getByText('bug'));

      expect(getByText('bug')).toHaveClass('is-danger');
      expect(window.location.search).toEqual('selectedTypes=bug');
    });

    it('toggles feature on click', () => {
      const { getByText } = renderSubject();

      fireEvent.click(getByText('feature'));

      expect(getByText('feature')).toHaveClass('is-primary');
      expect(window.location.search).toEqual('selectedTypes=feature');
    });

    it('toggles chore on click', () => {
      const { getByText } = renderSubject();

      fireEvent.click(getByText('chore'));

      expect(getByText('chore')).toHaveClass('is-info');
      expect(window.location.search).toEqual('selectedTypes=chore');
    });

    it('toggles blocked on click', () => {
      const { getByText } = renderSubject();

      fireEvent.click(getByText('blocked'));

      expect(getByText('blocked')).toHaveClass('is-warning');
      expect(window.location.search).toEqual('selectedTypes=blocked');
    });
  });
});
