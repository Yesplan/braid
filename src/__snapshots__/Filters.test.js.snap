// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters renders as per snapthot 1`] = `
<div
  class="media"
>
  <div
    class="media-left"
  >
    <h5
      class="subtitle is-5"
      style="line-height: 1.8;"
    >
      By owner
    </h5>
  </div>
  <div
    class="media-content"
  >
    <div
      class="field is-grouped is-grouped-multiline"
    >
      <p
        class="control"
      >
        <button
          class="button is-rounded "
          value="1"
        >
          <span
            class="is-uppercase"
          >
            AA
          </span>
        </button>
      </p>
      <p
        class="control"
      >
        <button
          class="button is-rounded "
          value="2"
        >
          <span
            class="is-uppercase"
          >
            BB
          </span>
        </button>
      </p>
      <p
        class="control"
      >
        <button
          class="button is-rounded "
          value="3"
        >
          <span
            class="is-uppercase"
          >
            CC
          </span>
        </button>
      </p>
      <div
        class="field has-addons"
        style="margin-bottom: .75rem;"
      >
        <p
          class="control"
        >
          <button
            class="button"
            data-testid="prevOwner"
          >
            <span
              class="icon"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-backward fa-w-16 "
                data-icon="backward"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11.5 280.6l192 160c20.6 17.2 52.5 2.8 52.5-24.6V96c0-27.4-31.9-41.8-52.5-24.6l-192 160c-15.3 12.8-15.3 36.4 0 49.2zm256 0l192 160c20.6 17.2 52.5 2.8 52.5-24.6V96c0-27.4-31.9-41.8-52.5-24.6l-192 160c-15.3 12.8-15.3 36.4 0 49.2z"
                  fill="currentColor"
                />
              </svg>
            </span>
            <span>
              <span
                class="is-underline"
              >
                P
              </span>
              rev
            </span>
          </button>
        </p>
        <p
          class="control"
        >
          <button
            class="button"
            data-testid="nextOwner"
          >
            <span
              class="icon"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-forward fa-w-16 "
                data-icon="forward"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M500.5 231.4l-192-160C287.9 54.3 256 68.6 256 96v320c0 27.4 31.9 41.8 52.5 24.6l192-160c15.3-12.8 15.3-36.4 0-49.2zm-256 0l-192-160C31.9 54.3 0 68.6 0 96v320c0 27.4 31.9 41.8 52.5 24.6l192-160c15.3-12.8 15.3-36.4 0-49.2z"
                  fill="currentColor"
                />
              </svg>
            </span>
            <span>
              <span
                class="is-underline"
              >
                N
              </span>
              ext
            </span>
          </button>
        </p>
        <p
          class="control"
        >
          <button
            class="button"
            data-testid="clearOwners"
            disabled=""
          >
            <span
              class="icon"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-times fa-w-11 "
                data-icon="times"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 352 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                  fill="currentColor"
                />
              </svg>
            </span>
            <span>
              <span
                class="is-underline"
              >
                C
              </span>
              lear
            </span>
          </button>
        </p>
      </div>
    </div>
  </div>
</div>
`;
